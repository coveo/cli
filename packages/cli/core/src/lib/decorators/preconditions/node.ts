import {getBinVersionPrecondition} from './binPreconditionsFactory.js';

export const IsNodeVersionInRange = getBinVersionPrecondition('node', {
  prettyName: 'Node.js',
});
