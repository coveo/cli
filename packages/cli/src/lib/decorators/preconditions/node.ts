import {constants} from 'os';
import * as dedent from 'dedent';

import type Command from '@oclif/command';
import {valid as isVersionValid, lt as isVersionLessThan} from 'semver';

import {SpawnProcessOutput, spawnProcessOutput} from '../../utils/process';

const linkToReadme =
  'https://github.com/coveo/cli/wiki/Node.js,-NPM-and-NPX-requirements';

export function IsNodeVersionAbove(requiredVersion: string) {
  return async function (target: Command) {
    if (!isVersionValid(requiredVersion)) {
      target.warn(dedent`
        Required version invalid: "${requiredVersion}".
        Please report this error to Coveo: https://github.com/coveo/cli/issues/new
      `);
      return false;
    }

    const output = await spawnProcessOutput('node', ['--version']);

    if (isMissingExecutable(output)) {
      target.warn(`${target.id} requires Node.js to run.`);
      warnHowToInstallNode(target);
      return false;
    }

    if (output.stderr || output.exitCode !== 0) {
      target.warn(dedent`
        ${target.id} requires a valid Node.js installation to run.
        An unknown error happened while trying to determine your node version with node --version
        ${output.stderr}
      `);
      warnHowToInstallNode(target);
      return false;
    }

    if (isVersionLessThan(output.stdout, requiredVersion)) {
      target.warn(dedent`
        ${target.id} needs a Node.js version greater than ${requiredVersion}
        Version detected: ${output.stdout}
      `);
      warnHowToInstallNode(target);
      return false;
    }

    return true;
  };
}

export function warnHowToInstallNode(target: Command) {
  target.warn(dedent`
    Please visit ${linkToReadme} for more detailed installation information.
  `);
}

export function isMissingExecutable(output: SpawnProcessOutput) {
  return (
    output.exitCode !== null &&
    Math.abs(output.exitCode) === constants.errno.ENOENT
  );
}
