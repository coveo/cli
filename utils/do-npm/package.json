{
  "name": "@coveo/do-npm",
  "version": "1.0.0",
  "private": true,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsc"
  },
  "devDependencies": {
    "@types/npmcli__arborist": "^5.6.1",
    "@types/npmcli__package-json": "^2.0.0",
    "@types/tmp": "^0.2.3",
    "@verdaccio/types": "^10.7.0",
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "@npmcli/arborist": "^6.2.0",
    "tmp": "^0.2.1",
    "npm": "^9.4.0"
  }
}
