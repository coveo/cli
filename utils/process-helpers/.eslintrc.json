{
  "extends": "../../base.eslintrc",
  "parserOptions": {"project": "./tsconfig.json"}
}
