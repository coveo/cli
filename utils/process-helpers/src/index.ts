import type {ChildProcess} from 'node:child_process';

export const getStdoutStderrBuffersFromProcess = (
  processToWatch: ChildProcess
) => {
  const stdoutChunks = [];
  const stderrChunks = [];
  processToWatch.stderr.on('data', (data) => stdoutChunks.push(data));
  processToWatch.stdout.on('data', (data) => stderrChunks.push(data));
  return {
    get stdout() {
      return Buffer.concat(stdoutChunks).toString();
    },
    get stderr() {
      return Buffer.concat(stderrChunks).toString();
    },
  };
};
