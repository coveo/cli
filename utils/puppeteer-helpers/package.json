{
  "name": "@coveo/puppeteer-helpers",
  "version": "1.0.0",
  "private": true,
  "type": "module",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    }
  },
  "scripts": {
    "build": "tsc"
  },
  "devDependencies": {
    "typescript": "^4.9.5"
  },
  "dependencies": {
    "puppeteer": "19.6.3"
  }
}
