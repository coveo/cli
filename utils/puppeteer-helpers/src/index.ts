import type {HTTPRequest, HTTPResponse} from 'puppeteer';

// TODO: Refactor when implementing into coveo/cli
// Copy from https://github.com/coveo/cli/blob/07b6c64ac76e6e508a44ad932dd9569d66561c9f/packages/cli-e2e/utils/platform.ts#L20-L21
export function isSearchRequestOrResponse(
  requestOrResponse: HTTPRequest | HTTPResponse
) {
  const searchUrl = new URL(
    '/rest/search/v2?organizationId',
    'https://platform.cloud.coveo.com'
  );
  return requestOrResponse.url().startsWith(searchUrl.href);
}
