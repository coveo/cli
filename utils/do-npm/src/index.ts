import {
  spawnSync,
  spawn,
  SpawnOptions,
  SpawnSyncOptions,
} from 'node:child_process';
import {dirname, join} from 'node:path';
import {createRequire} from 'node:module';

const require = createRequire(import.meta.url);
const npmCliPath = join(dirname(require.resolve('npm')), 'bin', 'npm-cli.js');

export const npmSync = (args: string[], opts?: SpawnSyncOptions) =>
  spawnSync('node', [npmCliPath, ...args], opts);
export const npmAsync = (args: string[], opts?: SpawnOptions) =>
  spawn('node', [npmCliPath, ...args], opts);
