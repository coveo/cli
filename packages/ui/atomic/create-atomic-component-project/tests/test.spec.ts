import {ChildProcess} from 'node:child_process';
import {join} from 'node:path';
import {mkdirSync} from 'node:fs';
import {lookup} from 'node:dns/promises';
import {npmSync, npmAsync} from '@coveo/do-npm';
import {getStdoutStderrBuffersFromProcess} from '@coveo/process-helpers';
import {isSearchRequestOrResponse} from '@coveo/puppeteer-helpers';
import {startVerdaccio} from '@coveo/verdaccio-starter';
import {Browser, HTTPRequest, launch as launchBrowser, Page} from 'puppeteer';
import {hashElement} from 'folder-hash';
import {DirResult, dirSync} from 'tmp';
import waitOn from 'wait-on';
import retry from 'async-retry';
import treeKill from 'tree-kill-promise';

const PACKAGE_NAME = '@coveo/create-atomic-component-project';

describe(PACKAGE_NAME, () => {
  let verdaccioProcess: ChildProcess;
  let verdaccioUrl: string;
  let tempDirectory: DirResult;
  let testDirectory: string;

  beforeAll(async () => {
    ({verdaccioUrl, verdaccioProcess} = await startVerdaccio(PACKAGE_NAME));
    tempDirectory = dirSync({unsafeCleanup: true});
    testDirectory = join(tempDirectory.name, 'testDirectory');
    mkdirSync(testDirectory);
  });

  afterAll(async () => {
    await treeKill(verdaccioProcess.pid);
    tempDirectory.removeCallback();
  });

  it('should initialize a base project', async () => {
    npmSync(['init', PACKAGE_NAME.replace('/create-', '/')], {
      env: {
        ...process.env,
        npm_config_registry: verdaccioUrl,
        npm_config_cache: dirSync({unsafeCleanup: true}).name,
      },
      cwd: testDirectory,
    });

    expect(
      await hashElement(testDirectory, {
        folders: {
          ignoreRootName: true,
          ignoreBasename: true,
        },
        files: {
          include: ['*'],
          ignoreRootName: true,
          ignoreBasename: true,
        },
      })
    ).toMatchSnapshot();
  });

  it('should be able to install all deps without issues', () => {
    expect(
      npmSync(['install'], {
        cwd: testDirectory,
        env: {...process.env, npm_config_registry: verdaccioUrl},
      }).status
    ).toBe(0);
  });

  describe('when running npm start', () => {
    const serveAddressMatcher = /http:\/\/localhost:\d*/gm;
    let serveAddress: string;
    let startProcess: ChildProcess;
    let browser: Browser;
    let page: Page;
    let interceptedRequests: HTTPRequest[];
    let startProcessOutputs: {
      readonly stdout: string;
      readonly stderr: string;
    };

    beforeAll(async () => {
      startProcess = npmAsync(['start'], {cwd: testDirectory});
      startProcessOutputs = getStdoutStderrBuffersFromProcess(startProcess);
      browser = await launchBrowser();
      page = await browser.newPage();
      interceptedRequests = [];
      page.on('request', (request: HTTPRequest) => {
        interceptedRequests.push(request);
      });
    });

    afterAll(async () => {
      await browser.close();
      await treeKill(startProcess.pid);
    });

    it('should eventually display the serve address', async () => {
      await retry(
        () => {
          expect(startProcessOutputs.stderr).toMatch(serveAddressMatcher);
        },
        {
          maxRetryTime: 2 * 60e3,
        }
      );
      const serveAddressUrl = new URL(
        serveAddressMatcher.exec(startProcessOutputs.stderr)[0]
      );
      const resolvedHostname = (await lookup(serveAddressUrl.hostname, 4))
        .address;
      serveAddressUrl.hostname = resolvedHostname;
      serveAddress = serveAddressUrl.toString();
    });

    it('should actually serve on this address', async () => {
      await expect(
        waitOn({
          resources: [serveAddress],
        })
      ).resolves.not.toThrow();
    });

    it('should build succesfully', async () => {
      expect(startProcessOutputs.stderr).not.toMatch(
        'build failed, watching for changes'
      );
      expect(startProcessOutputs.stderr).toMatch(
        'build finished, watching for changes'
      );
    });

    it('should do some search query when loading the page', async () => {
      await page.goto(serveAddress, {waitUntil: 'networkidle2'});
      expect(interceptedRequests.some(isSearchRequestOrResponse)).toBe(true);
    });
  });
});
