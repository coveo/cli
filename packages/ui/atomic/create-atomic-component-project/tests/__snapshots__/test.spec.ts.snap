// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@coveo/create-atomic-component-project should initialize a base project 1`] = `
HashedFolder {
  "children": [
    HashedFile {
      "hash": "/EUyu6gVZlO8FrKPYi4T6pPEZOI=",
      "name": "package.json",
    },
    HashedFile {
      "hash": "t2ijBtfSMeYJbqfftNaaKO/djIo=",
      "name": "readme.md",
    },
    HashedFolder {
      "children": [
        HashedFile {
          "hash": "YVuwXfIZfjfeVyLQ6cMy3kUqC3A=",
          "name": "html.d.ts",
        },
        HashedFolder {
          "children": [
            HashedFile {
              "hash": "wRrWASx36kB/WHP3xotlHQbcLdo=",
              "name": "index.css",
            },
            HashedFile {
              "hash": "PzWUKvDClNOd5HSK8c6YTFV8/9Q=",
              "name": "index.html",
            },
            HashedFile {
              "hash": "ZRD3OAMjlKnlu7Yh3iQv3gCcG8c=",
              "name": "index.ts",
            },
          ],
          "hash": "KaLKYWk10KFfD5NR/eQK0Zlk4qc=",
          "name": "pages",
        },
      ],
      "hash": "k4Tz5Ow35z6eAe5vRRIeg5Q7EQM=",
      "name": "src",
    },
    HashedFile {
      "hash": "BVZfiCzL+2kulE12K4HLAIl3kJ8=",
      "name": "stencil.config.ts",
    },
    HashedFile {
      "hash": "yMxJnSeOFw52ObOJrKQEh86v+HM=",
      "name": "tsconfig.json",
    },
  ],
  "hash": "Uk14EzUDjeEDKM690n+x3Zd2oVI=",
  "name": "testDirectory",
}
`;
